[domain-deletion]Introduce a feature flag to control domain deletion

gazi-yestemirova opened the pull request:
<!-- Describe what has changed in this PR -->
**What changed?**
This PR introduces a new feature flag,`EnableDomainDeletion`, to control access to the domain deletion API. 
This flag now gates the DeleteDomain API. If the flag is disabled, domain deletion requests will be rejected.
The flag defaults to false to prevent accidental domain deletions during rollout.

<!-- Tell your future self why have you made these changes -->
**Why?**
The primary goal is to make the domain deletion feature backward-compatible. We will enable the flag after the gradual rollout of the domain deletion API across all clusters.

<!-- How have you verified this change? Tested locally? Added a unit test? Checked in staging env? -->
**How did you test it?**
Unit tests & local testing

<!-- Assuming the worst case, what can be broken when deploying this change to production? -->
**Potential risks**
-

<!-- Is it notable for release? e.g. schema updates, configuration or data migration required? If so, please mention it, and also update CHANGELOG.md -->
**Release notes**

<!-- Is there any documentation updates should be made for config, https://cadenceworkflow.io/docs/operation-guide/setup/ ? If so, please open an PR in https://github.com/cadence-workflow/cadence-docs -->
**Documentation Changes**

dkrotx commented on service/frontend/api/domain_handlers.go:
```diff
@@ -154,6 +154,11 @@ func (wh *WorkflowHandler) DeleteDomain(ctx context.Context, deleteRequest *type
 	}
+
+	if !wh.config.EnableDomainDeletion() {
+		return &types.BadRequestError{Message: "Domain deletion is not enabled."}
```
nit: I'd change this to "Domain deletion feature is not enabled" so it is clear it is not something not enabled for the particular domain.

dkrotx replied:
At the same time what could be wrong with using this feature?
Is this allowed for a random persons?

3vilhamster replied:
It is related to the previous PR and the discussion https://github.com/cadence-workflow/cadence/pull/6918#pullrequestreview-2842043223
The problem is that we are introducing a new replication task requiring cross-cluster coordination.
@gazi-yestemirova Correct me if I'm wrong :)

gazi-yestemirova replied:
yes, that is correct

3vilhamster replied:
Could you add this to the PR description and point to the issue with replication tasks so it will be easier to understand why we need a feature flag in this case?

3vilhamster commented:
Could you elaborate on the backward-compatibility of the feature?
As far as I understand, we've just introduced a new feature, so why should it be guarded by the feature flag?

gazi-yestemirova commented:
> Could you elaborate on the backward-compatibility of the feature? As far as I understand, we've just introduced a new feature, so why should it be guarded by the feature flag?

When we delete a domain, if it is a global domain, we trigger a replication task to delete the domain in a replica cluster. 
I recently introduced IDL changes in the replicator, specifically DomainOperationDelete to be able to execute delete in a replica cluster.

Sometimes, it might happen so that Cluster A has these IDL changes but Cluster B is not deployed yet (or even when the rollbacks happen). Cluster A starts generating **delete tasks**, which get replicated to cluster B. When cluster B doesn't recognize it and falls into the default case and the task will be discarded. Nothing will crash, but we will end up with inconsistent data across active - passive clusters.

So this feature flag will be enabled when we have deployed the API changes and removed after the feature is rolled out across all the production environments.

dkrotx commented:
> So this feature flag will be enabled when we have deployed the API changes and removed after the feature is rolled out across all the production environments.

Does that mean you will deprecate this flag afterwards?

gazi-yestemirova commented:
> > So this feature flag will be enabled when we have deployed the API changes and removed after the feature is rolled out across all the production environments.
> 
> Does that mean you will deprecate this flag afterwards?

Yes, it will be removed. After I roll out the domain deletion feature to all the waves
